### PR TITLE
Handled IPAM policy properly to fully support strict IPAM

### PR DESCRIPTION
**Pull request by Ex4amp1e**
Issue: https://github.com/networkservicemesh/cmd-nse-icmp-responder/issues/614
Problem: strict IMAP has not been working for dual stack NSE
Solution: Place strict IPAM policy over group policy if it is configured by ENV

**How to test**
Setup NSE env
```
        - name: NSM_IPAM_POLICY
          value: strict
        - name: NSM_CIDR_PREFIX
          value: 172.16.1.100/27,2001:db8::/116
```

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -94,7 +92,7 @@ type Config struct {
 	ConnectTo              url.URL           `default:"unix:///var/lib/networkservicemesh/nsm.io.sock" desc:"url to connect to" split_words:"true"`
 	MaxTokenLifetime       time.Duration     `default:"10m" desc:"maximum lifetime of tokens" split_words:"true"`
 	RegistryClientPolicies []string          `default:"etc/nsm/opa/common/.*.rego,etc/nsm/opa/registry/.*.rego,etc/nsm/opa/client/.*.rego" desc:"paths to files and directories that contain registry client policies" split_words:"true"`
-	IPAMPolicy             ipamPolicyFunc    `default:"polite" desc:"defines NSE's IPAM Policy. Possible values: polite, strict. Polite policy accepts any addresses sent by client. Strict policy resets ip_context if any of the client's addresses doesn't match endpoint's CIDR." split_words:"true"`
+	IPAMPolicy             string            `default:"polite" desc:"defines NSE's IPAM Policy. Possible values: polite, strict. Polite policy accepts any addresses sent by client. Strict policy resets ip_context if any of the client's addresses doesn't match endpoint's CIDR." split_words:"true"`
```
Comment (denis-tingaikin) [inline main.go:95]:
```suggestion
	IPAMPolicy             newIPAMFunc            `default:"polite" desc:"defines NSE's IPAM Policy. Possible values: polite, strict. Polite policy accepts any addresses sent by client. Strict policy resets ip_context if any of the client's addresses doesn't match endpoint's CIDR." split_words:"true"`
```

File: main.go [hunk 2]
```diff
@@ -111,23 +109,6 @@ type Config struct {
 	PprofListenOn          string            `default:"localhost:6060" desc:"pprof URL to ListenAndServe" split_words:"true"`
 }
```
Comment (denis-tingaikin) [inline main.go:111]:
```suggestion
type newIPAMFunc func(...*net.IPNet) networkservice.NetworkServiceServer

// Decode takes a string IPAM Policy and returns the IPAM Policy func
func (f *newIPAMFunc) Decode(policy string) error {

	switch config.IPAMPolicy {
	case "strict":
	   *f = func(cidrPrefix [][]*net.IPNet) {
	   ipnetList := []*net.IPNet{}
	for _, group := range config.CidrPrefix {
		ipnetList = append(ipnetList, group...)
	}
		return strictipam.NewServer(func(i ...*net.IPNet) networkservice.NetworkServiceServer {
			return groupipam.NewServer(config.CidrPrefix)
		}, ipnetList...)
	case "polite":
			   *f = func(cidrPrefix [][]*net.IPNet) { 
			   return groupipam.NewServer(config.CidrPrefix)
			  }
	default:
		logrus.Fatalf("not a valid IPAM Policy: %s", config.IPAMPolicy)
	}
}
```

File: main.go [hunk 3]
```diff
@@ -239,13 +220,30 @@ func main() {
...
+	switch config.IPAMPolicy {
+	case "strict":
+		IPAMServer = strictipam.NewServer(func(i ...*net.IPNet) networkservice.NetworkServiceServer {
+			return groupipam.NewServer(config.CidrPrefix)
+		}, ipnetList...)
+	case "polite":
+		IPAMServer = groupipam.NewServer(config.CidrPrefix)
+	default:
+		logrus.Fatalf("not a valid IPAM Policy: %s", config.IPAMPolicy)
+	}
```
Comment (denis-tingaikin) [inline main.go:238]:
```suggestion
```

File: main.go [hunk 4]
```diff
@@ -239,13 +220,30 @@ func main() {
...
+
 	responderEndpoint := endpoint.NewServer(ctx,
 		spiffejwt.TokenGeneratorFunc(source, config.MaxTokenLifetime),
 		endpoint.WithName(config.Name),
 		endpoint.WithAuthorizeServer(authorize.NewServer()),
 		endpoint.WithAdditionalFunctionality(
 			onidle.NewServer(ctx, cancel, config.IdleTimeout),
-			groupipam.NewServer(config.CidrPrefix, groupipam.WithCustomIPAMServer(config.IPAMPolicy)),
+			IPAMServer,
```
Comment (denis-tingaikin) [inline main.go:246]:
```suggestion
			config.newIPAMFunc(config.CidrPrefix),
```

File: main.go (hunk 1 above)
Comment (Ex4amp1e) [reply main.go]: Done

File: main.go (hunk 2 above)
Comment (Ex4amp1e) [reply main.go:111]: Done

File: main.go (hunk 3 above)
Comment (Ex4amp1e) [reply main.go]: Done

File: main.go (hunk 4 above)
Comment (Ex4amp1e) [reply main.go]: Done

**Review by denis-tingaikin — APPROVED** [review]

**Pull request merged**